### DISCUSSION TITLE
Detecting syslog-ng stop in destination program

### DISCUSSION THREAD

**Discussion by petiepooo** [syslog-ng/syslog-ng#4099] [Q&A] [answered]
I'm trying to figure out how to store state on an external program run as a destination.  I checked for SIGTERM, SIGQUIT, and SIGHUP as well as EOF on stdin to try to determine when I'm about to be killed, but I came up empty.

How can I, as a program destination launched by syslog-ng OSE v3.37, determine that syslog-ng is exiting and about to shut me down?

**Comment by bazsi** [comment, marked as answer]
We are sending a SIGTERM to the entire process group. Try to follow
syslog-ng with strace to see it.

Also, we execute the child process through /bin/sh maybe the shell is
impacting signal delivery somehow.

Also, at the end the stdin of the process definitely closes, so you will
get end EOF for sure.

**Reply by petiepooo** [reply]
I suspect you're right about /bin/sh preventing signals to my process. I'm running a pipeline as the command, eg: program("my_script.py | another_repl", keep-alive(yes));  Because it's a pipeline, I cannot just exec the process so it replaces /bin/sh and receives signals.

I suppose I could rewrite it to run the other repl as a subprocess so I could use exec...  That would also allow me to handle its prompts/errors rather than just sending them to /dev/null.

FWIW, I watched for an EOF on stdin, but I think the parent /bin/sh was shutting me down before that, as I never received it.  Perhaps I was exiting on SIGPIPE; I don't think I tested for that.

**Reply by petiepooo** [reply]
How would you feel about a new feature giving users the option to avoid using the shell on a program() destination?

I saw a note about redefining /bin/sh as a symlink to /bin/bash if /bin/sh is giving someone troubles.  I'd love to see a new program() destination option like 'use-shell()' (type: yes or no, default: yes).  It would allow a user to specify bash without altering system files by using this config:
> program("bash -c \\" command --option setting | command2 \\" ", use-shell(no));

Even better, it would also allow one to avoid the overhead of a shell entirely by using this config:
> program("/path/to/elf-binary --option setting", use-shell(no));

**Reply by petiepooo** [reply]
One more follow-up.  I figured out how to avoid the parent shell that I believe is blocking signals.  If I'm right, I should be able to see the SIGTERM in my script and respond to it now.

The details, in case someone else encounters the same thing:

To simplify syntax, I am using a launcher script with the following contents:
> #!/bin/bash
exec > >(exec another_repl)
exec my_script.y

The key is using process substitution with '> >(...)' instead of a pipe so that another_repl is a child of my_script.py rather than a peer.

Although it's harder to parse, it could also be run without the launcher helper using this obtuse syntax:
> program("exec bash -c 'exec my_script.py > >(exec another_repl)'", keep-alive(yes));

**Comment by petiepooo** [comment]
Confirmed.  I can see SIGTERM now as the immediate child of syslog-ng (with no shell intercepting signals).

EDIT: I may have spoken too soon.  If I don't have keep-alive(yes) and I run 'syslog-ng-ctl reload' I see the SIGTERM in the child process before it's restarted.  If I do 'syslog-ng-ctl stop' I don't see either SIGTERM or EOF...  Might I see some other signal instead, perhaps SIGPIPE?

EDIT 2: ok, I do see the SIGTERM.  It's working as intended.  Thanks!
